File: Exam_Review/AI/Neural_Network.py
import numpy as np

# Simple Neural Network Computation


# implement the nn from slide 16 of n lecture
# assume that w is given as an np array and b is also given

def sigma(x):
    return 1/(1+np.exp(-x))

x = np.random.rand(4096)

# W is of size 1 x 4096 x 6
W = np.random.array(1, 4096, 6)

# b is of size 1 x 6
b = np.random.array(1, 6)

# z_1 = sigma(SUM W(1, j, 1) * x(j) + b(1, 1))

z = np.zeros((1, 6))

for k in range(6):
    # for j in range(4096):
    #     z[0, k] == W[0, j, k] * x[j]
    z[0, k] = np.dot(W[0, :, k], x)
    z[0, k] += b[0, k]

z = sigma(z)

# Extract Template for person at index 2

templ = []
for j in range(4096):
    templ.append(W[0, j, 2])
# templ is the thing we compute the dot product of x with

# Example questions 
#   Given a neural network, calculate the z
#   Perhaps might have to read a file with the values for W and b


# Neural Network Graph

class Node:
    def __init__(self, name, value=None):
        self.name = name
        self.value = value
        self. edges = []

    def add_edge(self, node, weight):
        self.edges.append({"node": node, "weight": weight})

    def __repr__(self):
        return f"Node({self.name}, {self.value})"   

import random
def create_nn_graph(input_dim, output_dim):
    for i in range(input_dim):
        node = Node(f"x{i}")
    for k in range(output_dim):
        node = Node(f"z{k}")
        z.append(node)
    for i in range(input_dim):
        for k in range(output_dim):
            w = random.random()
            x[i].add_edge(z[k], w)
            z[k].add_edge(x[i], w)

# "Need to know" slide 32

# Basically just compute all the outputs of the hidden layer
# And find the input that gives the highest output for the particular node
# ie input a picture of dogs would make the neuron corresponding to dogs go high

# To compute the hidden layer

x = np.random.rand(4096)

n_hid = 10
# W is of size 1 x 4096 x 6
W0 = np.random.array(1, 4096, n_hid)
W1 = np.random.array(1, n_hid, 6)


# b is of size 1 x 6
b0 = np.random.array(1, n_hid)
b1 = np.random.array(1, 6)
def nn(x, W0, W1, b0, b1, n_hid):
    hid = np.zeros((1, n_hid))
    z = np.zeros((1, 6))
    for k in range(n_hid):
        hid[0, k] = np.dot(W0[0, :, k], x)
        hid[0, k] += b0[0, k]

    for k in range(6):
        # for j in range(4096):
        #     z[0, k] == W[0, j, k] * x[j]
        z[0, k] = np.dot(W1[0, :, k], x)
        z[0, k] += b[0, k]

    return z, hid

# for hidden neuron 7, find the xs that activate it the most
neuron_7_acts = []

for i in range(10000):
    z, hid = nn(x[i], W0, W1, b0, b1, n_hid)
    neuron_7_acts.append((hid[0, 7], i))

neuron_7_acts.sort(reverse = True)
# the top 10 activations are for x's at indices [0][1], [1][1] ...
for i in range(10):
    print(f"x[{neuron_7_acts[i][1]}])")

# For graphs he could ask to use adjacency lists, matrices

File: Exam_Review/AI/Neural_Network.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>

#define INPUT_SIZE 4096
#define HIDDEN_SIZE 10
#define OUTPUT_SIZE 6
#define NUM_INPUTS 10000

// --------------------- Activation ---------------------
double sigmoid(double x) {
    return 1.0 / (1.0 + exp(-x));
}

// --------------------- Neural Network ---------------------
void nn(double x[INPUT_SIZE], double W0[INPUT_SIZE][HIDDEN_SIZE], double W1[HIDDEN_SIZE][OUTPUT_SIZE],
        double b0[HIDDEN_SIZE], double b1[OUTPUT_SIZE],
        double hid[HIDDEN_SIZE], double out[OUTPUT_SIZE]) {

    for (int i = 0; i < HIDDEN_SIZE; ++i) {
        hid[i] = b0[i];
        for (int j = 0; j < INPUT_SIZE; ++j)
            hid[i] += W0[j][i] * x[j];
        hid[i] = sigmoid(hid[i]);
    }

    for (int i = 0; i < OUTPUT_SIZE; ++i) {
        out[i] = b1[i];
        for (int j = 0; j < HIDDEN_SIZE; ++j)
            out[i] += W1[j][i] * hid[j];
        out[i] = sigmoid(out[i]);
    }
}

// --------------------- Graph Structures ---------------------
typedef struct Edge {
    int target;
    double weight;
    struct Edge* next;
} Edge;

typedef struct Node {
    char name[16];
    double value;
    Edge* edges;
} Node;

// Add edge from `from` to `to` with weight
void add_edge(Node* nodes, int from, int to, double weight) {
    Edge* edge = (Edge*)malloc(sizeof(Edge));
    edge->target = to;
    edge->weight = weight;
    edge->next = nodes[from].edges;
    nodes[from].edges = edge;
}

// --------------------- Main ---------------------
int main() {
    srand(time(NULL));

    // Neural network weights and biases
    double W0[INPUT_SIZE][HIDDEN_SIZE];
    double W1[HIDDEN_SIZE][OUTPUT_SIZE];
    double b0[HIDDEN_SIZE];
    double b1[OUTPUT_SIZE];

    // Random initialization
    for (int i = 0; i < INPUT_SIZE; ++i)
        for (int j = 0; j < HIDDEN_SIZE; ++j)
            W0[i][j] = (double)rand() / RAND_MAX;

    for (int i = 0; i < HIDDEN_SIZE; ++i)
        for (int j = 0; j < OUTPUT_SIZE; ++j)
            W1[i][j] = (double)rand() / RAND_MAX;

    for (int i = 0; i < HIDDEN_SIZE; ++i)
        b0[i] = (double)rand() / RAND_MAX;
    
    for (int i = 0; i < OUTPUT_SIZE; ++i)
        b1[i] = (double)rand() / RAND_MAX;

    // Build graph for visualization
    Node input_nodes[INPUT_SIZE];
    Node hidden_nodes[HIDDEN_SIZE];
    Node output_nodes[OUTPUT_SIZE];

    for (int i = 0; i < INPUT_SIZE; ++i) {
        sprintf(input_nodes[i].name, "x%d", i);
        input_nodes[i].value = 0;
        input_nodes[i].edges = NULL;
    }

    for (int i = 0; i < HIDDEN_SIZE; ++i) {
        sprintf(hidden_nodes[i].name, "h%d", i);
        hidden_nodes[i].value = 0;
        hidden_nodes[i].edges = NULL;
    }

    for (int i = 0; i < OUTPUT_SIZE; ++i) {
        sprintf(output_nodes[i].name, "z%d", i);
        output_nodes[i].value = 0;
        output_nodes[i].edges = NULL;
    }

    // Connect graph (input -> hidden)
    for (int i = 0; i < INPUT_SIZE; ++i)
        for (int j = 0; j < HIDDEN_SIZE; ++j)
            add_edge(input_nodes, i, j, W0[i][j]);

    // Connect graph (hidden -> output)
    for (int i = 0; i < HIDDEN_SIZE; ++i)
        for (int j = 0; j < OUTPUT_SIZE; ++j)
            add_edge(hidden_nodes, i, j, W1[i][j]);

    // Input set
    double X[NUM_INPUTS][INPUT_SIZE];
    for (int i = 0; i < NUM_INPUTS; ++i)
        for (int j = 0; j < INPUT_SIZE; ++j)
            X[i][j] = (double)rand() / RAND_MAX;

    // Track top 10 activations for hidden neuron 7
    double top_acts[10] = {0};
    int top_inds[10] = {0};

    for (int i = 0; i < NUM_INPUTS; ++i) {
        double hid[HIDDEN_SIZE], out[OUTPUT_SIZE];
        nn(X[i], W0, W1, b0, b1, hid, out);
        double a = hid[7];
        for (int j = 0; j < 10; ++j) {
            if (a > top_acts[j]) {
                for (int k = 9; k > j; --k) {
                    top_acts[k] = top_acts[k-1];
                    top_inds[k] = top_inds[k-1];
                }
                top_acts[j] = a;
                top_inds[j] = i;
                break;
            }
        }
    }

    printf("Top 10 inputs that activate hidden neuron 7:\n");
    for (int i = 0; i < 10; ++i)
        printf("x[%d] -> activation: %f\n", top_inds[i], top_acts[i]);

    return 0;
}
